src/map.rs

```diff
@@ -241,58 +252,89 @@ pub fn process_loaded_tile_maps(
+                    // these insertions should be limited to sprites referenced by objects
+                    let tile_width = tileset.tile_width as f32;
+                    let tile_height = tileset.tile_height as f32;
+                    let texture_width = image.width as f32;
+                    let texture_height = image.height as f32;
+                    let columns = (texture_width / tile_width).floor() as usize;
+                    let rows = (texture_height / tile_height).floor() as usize;
+
+                    let has_new = (0..(columns * rows) as u32).fold(false, |total, next| {
+                        total || !texture_atlas_map.contains_key(&(tileset.first_gid + next))
+                    });
+                    if !has_new { continue; }
+
+                    let atlas = TextureAtlas::from_grid(
+                        texture_handle.clone(),
+                        Vec2::new(tile_width, tile_height),
+                        columns,
+                        rows,
+                    );
+                    let atlas_handle = texture_atlases.add(atlas);
+                    for i in 0..(columns * rows) as u32 {
+                        if texture_atlas_map.contains_key(&(tileset.first_gid + i)) {
+                            continue;
+                        }
+                        // println!("insert: {}", tileset.first_gid + i);
+                        texture_atlas_map
+                            .insert(tileset.first_gid + i, atlas_handle.clone());
+                    }
+                } else if let Some(_) = tileset.tiles.first().and_then(|tile| tile.images.first()) {
+                    // Image collection tileset
+                    for tile in tileset.tiles.iter() {
+                        let tile_image = tile.images.first()
+                            .unwrap_or_else(|| panic!("missing image for tile {} in image collection tileset {}", tile.id, tileset.name));
+                        let tile_gid = tileset.first_gid + tile.id - 1;
+                        if texture_atlas_map.contains_key(&tile_gid) {
+                            continue;
                         }
+
+                        let texture_path = map
+                            .image_folder
+                            .join(tile_image.source.as_str());
+                        let texture_handle = asset_server.load(texture_path);
+                        materials_map.insert(
+                            tile_gid,
+                            materials.add(texture_handle.clone().into())
+                        );
+
+                        // Create an atlas with only this single image
+                        let atlas = TextureAtlas::from_grid(
+                            texture_handle.clone(),
+                            Vec2::new(tile_image.width as f32, tile_image.height as f32),
+                            1,
+                            1
+                        );
```

Comment:
To reduce the amount of churn needed outside of `map.rs`, we just create a one-image `TextureAtlas` for each individual tile image.

An alternate approach would be to use `TextureAtlasBuilder`, however I'm not sure about the performance implications of that (e.g. is it something that should be happening in a build process outside of the main game executable? does it actually make rendering faster? does it affect loading times?). However, that could be explored in a separate PR.

Reply:
Right, it could definitely be interesting to dynamically build the atlas, but indeed some people will be doing that in a build process, which case you'd want to look up the atlas + index by the tile image file name. Probably worth providing both options.